dbms/src/Flash/Executor/PipelineExecutorStatus.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Executor/PipelineExecutorStatus.h>
+#include <assert.h>
+
+namespace DB
+{
+ExecutionResult PipelineExecutorStatus::toExecutionResult()
+{
+    auto get_err_msg = getErrMsg();
+    return get_err_msg.empty()
+        ? ExecutionResult::success()
+        : ExecutionResult::fail(get_err_msg);
+}
+
+String PipelineExecutorStatus::getErrMsg()
+{
+    std::lock_guard lock(mu);
+    return err_msg;
+}
+
+void PipelineExecutorStatus::toError(String && err_msg_)
+{
+    {
+        std::lock_guard lock(mu);
+        if (!err_msg.empty())
+            return;
+        err_msg = err_msg_.empty() ? empty_err_msg : std::move(err_msg_);
+    }
+    cancel();
+}
+
+void PipelineExecutorStatus::wait()
+{
+    std::unique_lock lock(mu);
+    cv.wait(lock, [&] { return 0 == active_event_count; });
```

Comment:
why not record unfinished event count and check if canceled/error or all finished here?

Reply:
Because here we must wait for all Event/Task to end, even if it is cancel or error to avoid some objects (Hashmap of Join) being destructed before the mpp task is destructed.

Reply:
And using `active_event_cout`, we can avoid scheduling some event.
https://github.com/pingcap/tiflash/pull/6520/files#diff-1c711dc7f04d71dcb7dab05221c713d6b74482826be722aa2f7df1de00dd2728R83

Reply:
> And using `active_event_cout`, we can avoid scheduling some event when query is cancelled. https://github.com/pingcap/tiflash/pull/6520/files#diff-1c711dc7f04d71dcb7dab05221c713d6b74482826be722aa2f7df1de00dd2728R83

I think It's not conflict with using record unfinished event. The question came to me just because I try to think if we can remove the assumption that eventB should be scheduled before eventA finished when eventA is dependancy of eventB. However, I think current design looks all right to me.

Reply:
IMO If we record the unfinished event count, and wait for the unfinished event count be zore, it means that all event should should call `OnEventFinish`, this means that even if it is cancel, all events will be called.